site/content/blog/2024-05-21-introducing-the-grpc-addon/index.md

```diff
@@ -0,0 +1,98 @@
+---
+title: "Introducing the gRPC Add-on"
+summary: "Introducing a new add-on for viewing and attacking gRPC endpoints."
+images:
+- https://www.zaproxy.org/blog/2024-05-21-introducing-the-grpc-addon/images/0-blog-grpc-intro.png
+type: post
+tags:
+- blog
+- grpc
+date: "2024-05-21"
+authors:
+    - amit
+---
+[gRPC](https://grpc.io/) APIs have become increasingly popular for building efficient and robust microservices architectures. 
+However, testing these APIs effectively can be challenging, especially when dealing with Protobuf message encoding and decoding. 
+In this blog post, we'll explore a custom extension designed to streamline the process of testing gRPC endpoints and handling Protobuf messages.
+
+![ZAPbot attacking gRPC](images/0-blog-grpc-intro.png)
+
+The add-on is still in an early stage, so the range of its functionality is limited. 
+However, you can combine it with existing ZAP functionality to abuse gRPC endpoints in many different ways.
+
+This extension aims to simplify the testing process by providing tools for:
+
+* **Encoding and Decoding** : Easily encode and decode Protobuf messages for testing purposes.
+* **Testing gRPC Endpoints** : You can combine it with existing ZAP functionality to abuse gRPC endpoints in many different ways.
+
+### Installation
+
+If you are using the latest version of ZAP then you can download the gRPC add-on from the 
+[ZAP Marketplace](/addons/).
+
+### How to Display a gRPC Decoded Message
+
+![Displaying a gRPC decoded message](images/gRPCResponseBody.png)
+
+### Understanding Decoded Protobuf Messages
+
+When decoding Protobuf messages, it's essential to understand the structure of the decoded data. 
+Protobuf messages consist of fields, each with a field number, wire type, and value. 
+Let's delve deeper into these components:
+
+#### Field Structure
+
+* **Field Number** : Each field in a Protobuf message is assigned a unique field number. This number is used to identify and distinguish between different fields within the message.
+* **Wire Type** : The wire type determines how the field's value is encoded and provides information about the type of data stored in the field.
+* **Value** : The actual data contained within the field, encoded according to its wire type.
+
+### Protobuf Message Structure example
+
+Let's understand the decoded Protobuf Message Structure using an example.
+
+This is the Base64 encoded form of the binary text received by ZAP.
+
+```
+AAAAAIYJZCZMzMzcXkAVrseHQhi5YCDqrcDlJCixqAM1QEIPADOQSGVsbG8sIFByb3RvYnVmIUJMCgs
+xMjMgTWFpbiBTdBIGTX1DaXR5GgUxMjMONSIuCgtIZWxsbyBXb3JsZBIITXkgV29ybGQaCllvdXIgV2
+9ybGQiCU91ciBXb3JsZEjqrcDLJA==
+```
+
+On the left side, we have the original message retrieved by the gRPC endpoint in the backend. On the right side, we have the decoded message from the above encoded text, processed by the ZAP gRPC addon.
```

Comment:
In the text, it is written that we are comparing left and right image, but there is only one image, either we need to add original message image or we need to change this text.